sweep/sweeper.go

```diff
@@ -833,6 +834,103 @@ func (s *UtxoSweeper) clusterBySweepFeeRate(inputs pendingInputs) []inputCluster
 	return inputClusters
 }
 
+// zipClusters merges pairwise clusters from as and bs such that cluster a from
+// as is merged with a cluster from bs that has at least the fee rate of a.
+// This to ensure we don't delay confirmation by decreasing the fee rate (the
+// lock time inputs are typically second level HTLC transactions, that are time
+// sensitive).
+func zipClusters(as, bs []inputCluster) []inputCluster {
+	// Sort the clusters by decreasing fee rates.
+	sort.Slice(as, func(i, j int) bool {
+		return as[i].sweepFeeRate >
+			as[j].sweepFeeRate
+	})
+	sort.Slice(bs, func(i, j int) bool {
+		return bs[i].sweepFeeRate >
+			bs[j].sweepFeeRate
+	})
+
+	var (
+		finalClusters []inputCluster
+		j             int
+	)
+
+	// Go through each cluster in as, and merge with the next one from bs
+	// if it has at least the fee rate needed.
+	for i := range as {
+		a := as[i]
+
+		// Check the next one from bs if any.
+		if j < len(bs) {
+			b := bs[j]
+
+			// If the fee rate is less than a's, we break. All the
+			// remining clusters from bs have lower fee rate.
+			if b.sweepFeeRate < a.sweepFeeRate {
+				break
```

Comment:
Not sure if I'm missing something, but don't we need to include the remaining input clusters from `as` before the break? Otherwise they won't be returned as part of `finalClusters`.

Reply:
Confirmed with this diff: 
```
diff --git a/sweep/sweeper_test.go b/sweep/sweeper_test.go
index 0104903cc..f6123837e 100644
--- a/sweep/sweeper_test.go
+++ b/sweep/sweeper_test.go
@@ -1513,10 +1513,10 @@ func TestZipClusters(t *testing.T) {
 				createCluster(testInputsA, 5000),
 			},
 			bs: []inputCluster{
-				createCluster(testInputsB, 7000),
+				createCluster(testInputsB, 3000),
 			},
 			res: []inputCluster{
-				createCluster(testInputsC, 7000),
+				createCluster(testInputsC, 5000),
 			},
 		},
 		{
@@ -1558,8 +1558,9 @@ func TestZipClusters(t *testing.T) {
 	for _, test := range testCases {
 		zipped := zipClusters(test.as, test.bs)
 		if !reflect.DeepEqual(zipped, test.res) {
-			t.Fatalf("[%s] unexpected result: %v",
-				test.name, spew.Sdump(zipped))
+			t.Fatalf("[%s] unexpected result: got %v, want %v",
+				test.name, spew.Sdump(zipped),
+				spew.Sdump(test.res))
 		}
 	}
 }
```

Tests fail as only `bs` is added and items from `as` are omitted all together. The diff just makes the fee rate of `bs` lower than all of `as` so the break gets hit immediately. 

An even simpler scenario to consider if when `bs` is empty (we cluster by lock time and all the inputs have a lock time). With this logic, they'll all be dropped.

Reply:
Another way of stating is that at each step for `as`, it either should be:
  * merged in with `bs`
  * or added to the final cluster (non-merged) 

The break here means that the last statement can be skipped all together at times.

Reply:
Good catch! Yeah, this was wrong. I think it is a left over from when I initially had a double loop here.

Went with @Roasbeef's suggestion, and added added the failing test cases: https://github.com/lightningnetwork/lnd/pull/4750/commits/b28bc4d6010f404d924941c32a975e34a653a86d